FastSimulation/FastSimProducer/src/ParticleLooper.cc

```diff
@@ -0,0 +1,238 @@
+{
+    int simTrackIndex = simTracks_->size();
+    simTracks_->emplace_back(particle->pdgId(),
+                    particle->momentum(),
+                    particle->simVertexIndex(),
+                    particle->genParticleIndex());
+    simTracks_->back().setTrackId(simTrackIndex);
+    return simTrackIndex;
+}
+
+std::unique_ptr<fastsim::Particle> fastsim::ParticleLooper::nextGenParticle()
+{
+    // only consider particles that start in the beam pipe and end outside the beam pipe
+    // try to get the decay time from pythia
+    // use hepmc units
+    // make the link simtrack to simvertex
+    // try not to change the simvertex structure
+    // print a couple of ttbar events to undertand the simtrack structure? nah...
+    
+    for ( ; genParticleIterator_ != genParticleEnd_ ; ++genParticleIterator_,++genParticleIndex_ ) // loop over gen particles
+    {
+    	// some handy pointers and references
+    	const HepMC::GenParticle & particle = **genParticleIterator_;
+    	const HepMC::GenVertex * productionVertex = particle.production_vertex();
+    	const HepMC::GenVertex * endVertex = particle.end_vertex();
+
+        // skip incoming particles
+        if(!productionVertex){
+            continue;
+        }
+
+    	// particle must be produced within the beampipe
+    	if(productionVertex->position().perp2()*lengthUnitConversionFactor2_ > beamPipeRadius2_)
+    	{
+    	    continue;
+    	}
+    	
+    	// particle must not decay before it reaches the beam pipe
+    	if(endVertex && endVertex->position().perp2()*lengthUnitConversionFactor2_ < beamPipeRadius2_)
+    	{
+    	    continue;
+    	}
+    	// make the particle
+    	std::unique_ptr<Particle> newParticle(
+    	    new Particle(particle.pdg_id(),
+    			 math::XYZTLorentzVector(productionVertex->position().x()*lengthUnitConversionFactor_,
+    						 productionVertex->position().y()*lengthUnitConversionFactor_,
+    						 productionVertex->position().z()*lengthUnitConversionFactor_,
+    						 productionVertex->position().t()*timeUnitConversionFactor_),
+    			 math::XYZTLorentzVector(particle.momentum().x()*momentumUnitConversionFactor_,
+    						 particle.momentum().y()*momentumUnitConversionFactor_,
+    						 particle.momentum().z()*momentumUnitConversionFactor_,
+    						 particle.momentum().e()*momentumUnitConversionFactor_)));
+    	newParticle->setGenParticleIndex(genParticleIndex_);
+
+    	// try to get the life time of the particle from the genEvent
+    	if(endVertex)
+    	{
+    	    double labFrameLifeTime = (endVertex->position().t() - productionVertex->position().t())*timeUnitConversionFactor_;
+    	    newParticle->setRemainingProperLifeTime(labFrameLifeTime * newParticle->gamma());
+    	}
+
+        // TODO: THIS HAS TO BE FIXED! E.g. the products of a b-decay should point to that vertex and not to the primary vertex!
+        newParticle->setSimVertexIndex(0);
```

Comment:
The following structure has to be used (according to Lukas):
Each simVertex has to have a parentSimTrackIndex, so the index of the particle/simTrack that produced the vertex.
Each simTrack/particle has to have a parentSimVertexIndex, so the index of the vertex it was produced.
Simple example: muGun
There is 1 simVertex which has the simTrackIndex -1 since there is no incoming track.
There are two tracks (the two muons), each of them has the simVertexIndex 0, so pointing to the primary vertex.
In case there are e.g. decay products of a b-quark already in the GEN event you'd somehow have to get the simTrack/simVertex structure from the GEN event and setting all indices accordingly. Unfortunately, I have no idea how to do this. I'll try to dig into the old code how it was done there.